### ISSUE TITLE
clij2-fft not working on arm64 mac

### ISSUE THREAD

**Issue by thawn** [BiAPoL/PoL-BioImage-Analysis-TS-GPU-Accelerated-Image-Analysis#9]
When I try to run clij2-fft on my arm64 mac, I get an error:

`OSError: dlopen(libclij2fft.dylib, [...] (no such file)`
<details>
get lib
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
Cell In[7], line 2
      1 if clij2_rl_found==True:
----> 2     decon_clij2=richardson_lucy_nc(im,psf,100,0)
      3     decon_helper.show_xyz_slice(decon_clij2,'tnia')
      4 else:

File [/opt/homebrew/Caskroom/mambaforge/base/envs/deconvolution-napari-env/lib/python3.9/site-packages/clij2fft/richardson_lucy.py:125](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/Caskroom/mambaforge/base/envs/deconvolution-napari-env/lib/python3.9/site-packages/clij2fft/richardson_lucy.py:125), in richardson_lucy_nc(img, psf, numiterations, regularizationfactor, lib)
    123 if (lib==None):
    124     print('get lib')
--> 125     lib = getlib()
    127 # the normalization factor is the valid region correlated with the PSF
    128 lib.convcorr3d_32f(int(normal.shape[2]), int(normal.shape[1]), int(normal.shape[0]), valid, shifted_psf, normal,1)

File [/opt/homebrew/Caskroom/mambaforge/base/envs/deconvolution-napari-env/lib/python3.9/site-packages/clij2fft/libs.py:26](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/Caskroom/mambaforge/base/envs/deconvolution-napari-env/lib/python3.9/site-packages/clij2fft/libs.py:26), in getlib()
     24     clij2fft=CDLL('libclij2fft.so', mode=RTLD_GLOBAL)
     25 elif os.name=='posix' and sys.platform =='darwin':
---> 26     clij2fft=CDLL('libclij2fft.dylib', mode=RTLD_GLOBAL)
     27 # if not posix assume windows
     28 else:
     29     clij2fft=CDLL('clij2fft.dll', mode=RTLD_GLOBAL)

File [/opt/homebrew/Caskroom/mambaforge/base/envs/deconvolution-napari-env/lib/python3.9/ctypes/__init__.py:374](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/Caskroom/mambaforge/base/envs/deconvolution-napari-env/lib/python3.9/ctypes/__init__.py:374), in CDLL.__init__(self, name, mode, handle, use_errno, use_last_error, winmode)
    371 self._FuncPtr = _FuncPtr
    373 if handle is None:
--> 374     self._handle = _dlopen(self._name, mode)
    375 else:
    376     self._handle = handle

OSError: dlopen(libclij2fft.dylib, 0x000A): tried: '/opt/homebrew/Caskroom/mambaforge/base/envs/deconvolution-napari-env/lib/python3.9/lib-dynload/../../libclij2fft.dylib' (no such file), '/opt/homebrew/Caskroom/mambaforge/base/envs/deconvolution-napari-env/bin/../lib/libclij2fft.dylib' (no such file), 'libclij2fft.dylib' (no such file), '/usr/local/lib/libclij2fft.dylib' (no such file), '/usr/lib/libclij2fft.dylib' (no such file), '/Users/korten/Documents/workspace/bnorthan/PoL-BioImage-Analysis-TS-GPU-Accelerated-Image-Analysis/docs/30_Deconvolution/libclij2fft.dylib' (no such file)</details>

I then tried to copy `libclij2fft.dylib` from my fiji installation, but then I get an error about a wrong architecture (x86_64 instead of arm64)

I found [on image.sc](https://forum.image.sc/t/fiji-clij-etc-native-on-apple-silicon-arm64-m1/53627) that there should be an arm64 version of clij2, but I failed to install fiji.app (I think it fails, because I am using openjdk instead of Azul)

@haesleinhuepf can I download the arm64  dylib files somewhere?

**Comment by bnorthan** [comment]
Hi @thawn 

Did you get the latest version?  Version 0.15 does not come with arm64 lib but version 0.16 (created just a couple of days ago) should, (and the Fijij version probably has a x86_64 lib too).  Peter Sobolewski helped me compile mac arm64 version last week.  If you have version 0.16 and it's still working let me know.  

If you want to grab the libraries directly the Arm64 clij2-fft and clFFT are [here](https://github.com/clij/clij2-fft/tree/master/lib/macosx-arm64) (though theoretically you should be able to get them now with the experimental pip install (pip install --index-url https://test.pypi.org/simple/ --no-deps clij2-fft)

**Comment by thawn** [comment]
Hi @bnorthan 

I just reinstalled and now the error message has changed:
```
OSError: dlopen(libclij2fft.dylib, 0x000A): Library not loaded: '/opt/homebrew/opt/clfft/lib/libclFFT.2.dylib'
  Referenced from: '/opt/homebrew/Caskroom/mambaforge/base/envs/deconvolution-napari-env/lib/libclij2fft.dylib'
  Reason: tried: '/opt/homebrew/opt/clfft/lib/libclFFT.2.dylib' (no such file), '/usr/local/lib/libclFFT.2.dylib' (no such file), '/usr/lib/libclFFT.2.dylib' (no such file)
```
I guess it should be looking for `/opt/homebrew/Caskroom/mambaforge/base/envs/deconvolution-napari-env/lib/libclFFT.2.dylib` but there is only `libclFFT.dylib` present there.
I tried to create a symlink for `libclFFT.2.dylib` pointing to `libclFFT.dylib`, but no luck.

Then I installed clfft via homebrew (`brew install clfft`) and now it works :-)

**Comment by thawn** [comment]
do you think it might be possible to provide `libclFFT.2.dylib` with `clij2-fft`?

**Comment by bnorthan** [comment]
Yes.  In fact it is already in the repo [here](https://github.com/clij/clij2-fft/tree/master/lib/macosx-arm64)

What we are doing is creating a universal lib with both the arm and x86_64 version then distributing that on PyPi

Command to create the universal lib is as follows. 

```
lipo -create -output lib/macosx-universal2/libclFFT.dylib lib/macosx/libclFFT.dylib lib/macosx-arm64/libclFFT.dylib
```

After we create the universal lib it is copied to a location under the python directory then added as a data_file in [setup.py](https://github.com/clij/clij2-fft/blob/master/python/setup.py)

After installation under your current environment libclij2fft.dylib and libclFFT.dylib should then exist under the 'lib' directory (along with un-needed linux versions, the build is a work in progress and some parts still need to be cleaned up). 

First thing to check is if the universal lib ends up in the right spot, and whether it can be found within the environment. 

Next thing I am wondering if we just need to make a universal lib for libclFFT.2.dylib too.  However I also see there is a 'libclFFT.2.12.2.dylib.

And the strange thing is that they are all the same size (900 kB or so).  So I don't think the libclFFT.dylib or libclFFT.2.dylib are symlinks.  

That being said it looks like on Linux I am only distributing libclFFT.so.2.  So maybe just need to make a universal lib of that one, and replace the other. 

I don't have a Mac to test on but if are you able to quickly check on this it would be helpful.  I'm also happy to touch base tomorrow and do a quick zoom to trouble shoot.

**Comment by bnorthan** [comment]
What about 9:30/3:30?

**Comment by thawn** [comment]
sounds great :-) I'll send a zoom link via email.

**Comment by thawn** [comment]
just tested it and it is fixed in clij2-fft v0.19. Thanks a lot for the fast fix @bnorthan ! :-)
